File: test/oxc/scope_test.rb
# frozen_string_literal: true

require "test_helper"

module Oxc
  class ScopeTest < Minitest::Spec
    SOURCE = <<~JS
      let count = 0
      const label = "Send"
      function bump() { const step = 1; count += step; render(label) }
      bump()
    JS

    SCOPED = %(let count_1a2b3c4d = 0;\nconst label_1a2b3c4d = "Send";\nfunction bump_1a2b3c4d() {\n\tconst step = 1;\n\tcount_1a2b3c4d += step;\n\trender(label_1a2b3c4d);\n}\nbump_1a2b3c4d();\n)

    test "renames what the file declared at the top level, and every reference to it" do
      assert_equal SCOPED, Oxc.scope(SOURCE, scope: "1a2b3c4d").code
    end

    test "leaves a name declared inside a scope of its own alone, and one the file never declared" do
      code = Oxc.scope(SOURCE, scope: "1a2b3c4d").code

      assert_equal "\tconst step = 1;\n", code.lines[3]
      assert_equal "\trender(label_1a2b3c4d);\n", code.lines[5]
    end

    test "says what each name became" do
      renamed = Oxc.scope(SOURCE, scope: "1a2b3c4d").renamed

      assert_equal({ "bump" => "bump_1a2b3c4d", "count" => "count_1a2b3c4d", "label" => "label_1a2b3c4d" }, renamed)
    end

    test "reads a separator of its own" do
      code = Oxc.scope("let a = 1; foo(a)", scope: "x", separator: "$").code

      assert_equal "let a$x = 1;\nfoo(a$x);\n", code
    end

    test "refuses a scope that could not read as part of a name" do
      error = assert_raises(OptionError) { Oxc.scope("let a = 1", scope: "1a-2b") }

      assert_equal 'Invalid scope "1a-2b". It has to read as part of a JavaScript name, so letters, digits, _ and $ only.',
                   error.message
    end

    test "refuses no scope at all" do
      error = assert_raises(OptionError) { Oxc.scope("let a = 1", scope: "") }

      assert_equal "scope has to be given, and has to read as part of a JavaScript name.", error.message
    end

    test "answers what it could not read instead of renaming it" do
      error = assert_raises(SyntaxError) { Oxc.scope("const x = ;", scope: "x") }

      assert_equal "Unexpected token", error.message
    end

    test "is a result of its own" do
      result = Oxc.scope("let a = 1; foo(a)", scope: "x")

      assert_kind_of Oxc::Result, result
      assert_instance_of Oxc::ScopeResult, result
      assert_predicate result, :frozen?
    end
  end
end

File: rust/tests/scope_test.rs
use oxc_ffi::{oxc_result_free, oxc_scope, OxcErrorCode};

use std::ffi::{CStr, CString};

fn scope(source: &str, options: &str) -> (OxcErrorCode, String) {
  let source = CString::new(source).unwrap();
  let options = CString::new(options).unwrap();

  let result = unsafe { oxc_scope(source.as_ptr(), options.as_ptr()) };
  let code = result.code;

  let answer = if result.value.is_null() {
    unsafe { CStr::from_ptr(result.error) }.to_str().unwrap().to_string()
  } else {
    unsafe { CStr::from_ptr(result.value) }.to_str().unwrap().to_string()
  };

  unsafe { oxc_result_free(result) };

  (code, answer)
}

#[test]
fn scope_renames_what_the_file_declared_at_the_top_level() {
  let (code, answer) = scope("let a = 1; foo(a)", r#"{"scope":"x"}"#);

  assert_eq!(code, OxcErrorCode::None);
  assert_eq!(
    answer,
    r#"{"code":"let a_x = 1;\nfoo(a_x);\n","renamed":{"a":"a_x"},"errors":[],"panicked":false}"#
  );
}

#[test]
fn scope_refuses_a_scope_that_could_not_read_as_part_of_a_name() {
  let (code, answer) = scope("let a = 1", r#"{"scope":"a-b"}"#);

  assert_eq!(code, OxcErrorCode::Option);
  assert_eq!(
    answer,
    concat!(
      r#"Invalid scope "a-b". It has to read as part of a JavaScript name, "#,
      "so letters, digits, _ and $ only."
    )
  );
}
